File: driftwatch/backend/internal/agent/diff.go
package agent

import (
	"sort"
	"strings"
	"time"

	"github.com/google/uuid"

	"github.com/souvik-biswas-dev/driftwatch/internal/docker"
)

const (
	DriftTypeEnvMismatch      = "env_mismatch"
	DriftTypeImageStale       = "image_stale"
	DriftTypePortChanged      = "port_changed"
	DriftTypeMissingContainer = "missing_container"
	DriftTypeExtraContainer   = "extra_container"
)

const (
	SeverityCritical = "critical"
	SeverityWarning  = "warning"
	SeverityInfo     = "info"
)

type DriftEvent struct {
	ID            string
	Type          string
	ContainerName string
	LiveValue     string
	DeclaredValue string
	Severity      string
	DetectedAt    time.Time
}

// Diff compares the live and declared snapshots and returns every drift
// event detected. The output ordering is stable: containers are walked in
// alphabetical order, and per-container env keys are walked in sorted order.
func Diff(live, declared *docker.LiveSnapshot) []DriftEvent {
	liveByName := indexByName(live)
	declaredByName := indexByName(declared)

	names := unionNames(liveByName, declaredByName)
	now := time.Now().UTC()

	var events []DriftEvent
	for _, name := range names {
		liveC, inLive := liveByName[name]
		decC, inDeclared := declaredByName[name]

		switch {
		case inDeclared && !inLive:
			events = append(events, DriftEvent{
				ID:            uuid.NewString(),
				Type:          DriftTypeMissingContainer,
				ContainerName: name,
				DeclaredValue: decC.Image,
				Severity:      SeverityCritical,
				DetectedAt:    now,
			})
			continue
		case inLive && !inDeclared:
			events = append(events, DriftEvent{
				ID:            uuid.NewString(),
				Type:          DriftTypeExtraContainer,
				ContainerName: name,
				LiveValue:     liveC.Image,
				Severity:      SeverityInfo,
				DetectedAt:    now,
			})
			continue
		}

		// Image tag — split on ":" and compare the tag half.
		if tagOf(liveC.Image) != tagOf(decC.Image) {
			events = append(events, DriftEvent{
				ID:            uuid.NewString(),
				Type:          DriftTypeImageStale,
				ContainerName: name,
				LiveValue:     liveC.Image,
				DeclaredValue: decC.Image,
				Severity:      SeverityWarning,
				DetectedAt:    now,
			})
		}

		// Env — for every declared key, flag if live missing or differs.
		envKeys := make([]string, 0, len(decC.Env))
		for k := range decC.Env {
			envKeys = append(envKeys, k)
		}
		sort.Strings(envKeys)
		for _, k := range envKeys {
			decVal := decC.Env[k]
			liveVal, ok := liveC.Env[k]
			if !ok || liveVal != decVal {
				events = append(events, DriftEvent{
					ID:            uuid.NewString(),
					Type:          DriftTypeEnvMismatch,
					ContainerName: name,
					LiveValue:     liveVal,
					DeclaredValue: decVal,
					Severity:      severityForEnvKey(k),
					DetectedAt:    now,
				})
			}
		}

		// Ports — declared set must be a subset of live set.
		liveSet := make(map[string]struct{}, len(liveC.Ports))
		for _, p := range liveC.Ports {
			liveSet[p] = struct{}{}
		}
		missing := false
		for _, p := range decC.Ports {
			if _, ok := liveSet[p]; !ok {
				missing = true
				break
			}
		}
		if missing {
			events = append(events, DriftEvent{
				ID:            uuid.NewString(),
				Type:          DriftTypePortChanged,
				ContainerName: name,
				LiveValue:     strings.Join(liveC.Ports, ","),
				DeclaredValue: strings.Join(decC.Ports, ","),
				Severity:      SeverityWarning,
				DetectedAt:    now,
			})
		}
	}

	return events
}

func indexByName(s *docker.LiveSnapshot) map[string]docker.ContainerState {
	out := make(map[string]docker.ContainerState)
	if s == nil {
		return out
	}
	for _, c := range s.Containers {
		out[c.Name] = c
	}
	return out
}

func unionNames(a, b map[string]docker.ContainerState) []string {
	set := make(map[string]struct{}, len(a)+len(b))
	for n := range a {
		set[n] = struct{}{}
	}
	for n := range b {
		set[n] = struct{}{}
	}
	out := make([]string, 0, len(set))
	for n := range set {
		out = append(out, n)
	}
	sort.Strings(out)
	return out
}

// tagOf returns the tag portion of a docker image reference. It splits on
// the last ":" so registries with a port ("registry:5000/img:tag") work,
// and falls back to "latest" when no tag is present.
func tagOf(image string) string {
	idx := strings.LastIndex(image, ":")
	if idx == -1 {
		return "latest"
	}
	return image[idx+1:]
}

func severityForEnvKey(key string) string {
	up := strings.ToUpper(key)
	for _, trigger := range []string{"PASSWORD", "SECRET", "KEY", "TOKEN", "DATABASE", "DB"} {
		if strings.Contains(up, trigger) {
			return SeverityCritical
		}
	}
	return SeverityWarning
}

File: driftwatch/backend/src/services/diff.ts
import { randomUUID } from 'node:crypto';

import {
	DriftType,
	Severity,
	type ContainerState,
	type DriftEvent,
	type LiveSnapshot,
	type SeverityValue
} from '../types.js';

/**
 * diff compares the live and declared snapshots and returns every drift event
 * detected. The output ordering is stable: containers are walked in
 * alphabetical order, and per-container env keys are walked in sorted order.
 */
export function diff(
	live: LiveSnapshot | null | undefined,
	declared: LiveSnapshot | null | undefined
): DriftEvent[] {
	const liveByName = indexByName(live);
	const declaredByName = indexByName(declared);
	const now = new Date();

	const events: DriftEvent[] = [];

	for (const name of unionNames(liveByName, declaredByName)) {
		const liveC = liveByName.get(name);
		const decC = declaredByName.get(name);

		if (decC && !liveC) {
			events.push({
				id: randomUUID(),
				type: DriftType.MissingContainer,
				containerName: name,
				liveValue: '',
				declaredValue: decC.image,
				severity: Severity.Critical,
				detectedAt: now
			});
			continue;
		}
		if (liveC && !decC) {
			events.push({
				id: randomUUID(),
				type: DriftType.ExtraContainer,
				containerName: name,
				liveValue: liveC.image,
				declaredValue: '',
				severity: Severity.Info,
				detectedAt: now
			});
			continue;
		}
		if (!liveC || !decC) continue; // unreachable; narrows the types below

		// Image tag — split on the last ":" and compare the tag half.
		if (tagOf(liveC.image) !== tagOf(decC.image)) {
			events.push({
				id: randomUUID(),
				type: DriftType.ImageStale,
				containerName: name,
				liveValue: liveC.image,
				declaredValue: decC.image,
				severity: Severity.Warning,
				detectedAt: now
			});
		}

		// Env — for every declared key, flag if live is missing it or differs.
		for (const k of Object.keys(decC.env).sort()) {
			const decVal = decC.env[k] ?? '';
			const liveVal = liveC.env[k];
			if (liveVal === undefined || liveVal !== decVal) {
				events.push({
					id: randomUUID(),
					type: DriftType.EnvMismatch,
					containerName: name,
					liveValue: liveVal ?? '',
					declaredValue: decVal,
					severity: severityForEnvKey(k),
					detectedAt: now
				});
			}
		}

		// Ports — the declared set must be a subset of the live set.
		const liveSet = new Set(liveC.ports);
		if (decC.ports.some((p) => !liveSet.has(p))) {
			events.push({
				id: randomUUID(),
				type: DriftType.PortChanged,
				containerName: name,
				liveValue: liveC.ports.join(','),
				declaredValue: decC.ports.join(','),
				severity: Severity.Warning,
				detectedAt: now
			});
		}
	}

	return events;
}

function indexByName(s: LiveSnapshot | null | undefined): Map<string, ContainerState> {
	const out = new Map<string, ContainerState>();
	if (!s?.containers) return out;
	for (const c of s.containers) out.set(c.name, c);
	return out;
}

function unionNames(
	a: Map<string, ContainerState>,
	b: Map<string, ContainerState>
): string[] {
	return [...new Set([...a.keys(), ...b.keys()])].sort();
}

/**
 * tagOf returns the tag portion of a docker image reference. It splits on the
 * last ":" so registries with a port ("registry:5000/img:tag") work, and falls
 * back to "latest" when no tag is present.
 */
export function tagOf(image: string): string {
	const idx = image.lastIndexOf(':');
	if (idx === -1) return 'latest';
	return image.slice(idx + 1);
}

const CRITICAL_ENV_TRIGGERS = ['PASSWORD', 'SECRET', 'KEY', 'TOKEN', 'DATABASE', 'DB'];

export function severityForEnvKey(key: string): SeverityValue {
	const up = key.toUpperCase();
	return CRITICAL_ENV_TRIGGERS.some((t) => up.includes(t))
		? Severity.Critical
		: Severity.Warning;
}
